pkg/collector/osmlogs_collector.go

```diff
@@ -0,0 +1,136 @@
+			fmt.Printf("Failed to collect OSM controller logs for mesh %s: %+v", meshName, err)
+		}
+	}
+
+	return nil
+}
+
+// Helper function to collect output of a given kubectl command to a file
+func collectResourceToFile(collector *OsmLogsCollector, rootPath, fileName string, kubeCmds []string) error {
+	resourceFile := filepath.Join(rootPath, fileName)
+	output, err := utils.RunCommandOnContainer("kubectl", kubeCmds...)
+	if err != nil {
+		return err
+	}
+
+	if err = utils.WriteToFile(resourceFile, output); err != nil {
+		return err
+	}
+
+	collector.AddToCollectorFiles(resourceFile)
+
+	return nil
+}
+
+// ** Collect data for namespaces monitored by a given mesh **
+func collectDataFromNamespaces(collector *OsmLogsCollector, namespaces []string, rootPath, meshName string) error {
+	for _, namespace := range namespaces {
+		var namespaceResourcesMap = map[string][]string{
+			meshName + "_" + namespace + "_metadata":          []string{"get", "namespaces", namespace, "-o=jsonpath={..metadata}", "-o=json"},
+			meshName + "_" + namespace + "_services_table":    []string{"get", "services", "-n", namespace},
+			meshName + "_" + namespace + "_services_configs":  []string{"get", "services", "-n", namespace, "-o", "json"},
+			meshName + "_" + namespace + "_endpoints_table":   []string{"get", "endpoints", "-n", namespace},
+			meshName + "_" + namespace + "_endpoints_configs": []string{"get", "endpoints", "-n", namespace, "-o", "json"},
+		}
+
+		for fileName, kubeCmd := range namespaceResourcesMap {
+			if err := collectResourceToFile(collector, rootPath, fileName, kubeCmd); err != nil {
+				fmt.Printf("Failed to collect %s in OSM monitored namespace %s: %+v", fileName, namespace, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ** Collect logs of every OSM controller in a given mesh **
+func collectControllerLogs(collector *OsmLogsCollector, rootPath, meshName string) error {
+	controllerInfos, err := getResourceList("pods", "app=osm-controller", "-o=custom-columns=NAME:{..metadata.name},NAMESPACE:{..metadata.namespace}", "\n")
+	if err != nil {
+		return err
+	}
+	for _, controllerInfo := range controllerInfos[1:] {
+		controllerInfoParts := strings.Fields(controllerInfo)
+		if len(controllerInfoParts) > 0 {
+			podName := controllerInfoParts[0]
+			namespace := controllerInfoParts[1]
+			if err := collectResourceToFile(collector, rootPath, meshName+"_controller_logs_"+podName, []string{"logs", "-n", namespace, podName}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Helper function to get all resoures of given type in the cluster
+func getResourceList(resource, label, outputFormat, separator string) ([]string, error) {
+	resourceList, err := utils.RunCommandOnContainer("kubectl", "get", resource, "--all-namespaces", "--selector", label, outputFormat)
+	if err != nil {
+		return nil, err
+	}
+	return strings.Split(strings.Trim(resourceList, "\""), separator), nil
+}
```

Comment:
I really like the idea of this chainable/hierarchical approach and +1 on defining values in the yaml... leaving specific commands in the code made us uncomfortable for sure. The reason we haven't yet been able to move these out into a yaml is that some of the values in most commands are variable (eg: they use the meshName or specific namespace). But with some more work to template resource collection I think it should be possible!

Reply:
Would the goal be to make this general for other collectors as well?

Reply:
Yep! Your current approach makes sense, first step get down a MVP / put requirements into code then can work on making it more generalizable. Probably wouldn't try make all collectors chainable but define a new IChainableCollector interface and those collectors for which it makes sense can implement (e.g. probably not nodeLogs collector).

I'm now thinking it might make more sense if the proposed "resourceCollector" is implemented via extension of the capabilities of the current "KubeObjects" collector instead, as these are equivalent in intent.

I'm leaning towards not committing to chainable collector functionality right now as there is a lot to work out there, but I think we can quite easily refactor what you have to push the specifics out to the yaml, and add a new general ServiceMeshCollector to consume it, I had a quick go at moving the specifics out to the yaml and this is what I have (apologies for formatting, and I think its probably not valid yaml but hopefully you get the idea):

---
apiVersion: v1 
kind: ConfigMap 
metadata:
  name: kubeobjects-config
  namespace: aks-periscope
data:
  DIAGNOSTIC_KUBEOBJECTS_LIST: kube-system/pod kube-system/service kube-system/deployment #periscope default
  
  diagnostic-kubeobjects.properties |                                        # add support for consuming this kubeobject definition format in kubeobject collector
    all-resources-table.kubeobject=all
    all-resources-table.namespace=all
    all-resources-table.selector=app.kubernetes.io/name=openservicemesh.io
    all-resources-table.output=wide
    
    all-resources-configs.kubeobject=all
    all-resources-configs.namespace=all
    all-resources-configs.selector=app.kubernetes.io/name=openservicemesh.io
    all-resources-configs.output=json
    
    mutating-webhook-configurations.kubeobject=MutatingWebhookConfiguration
    mutating-webhook-configurations.namespace=all
    mutating-webhook-configurations.selector=app.kubernetes.io/name=openservicemesh.io
    mutating-webhook-configurations.output=json
    
    validating-webhook-configurations.kubeobject=all
    validating-webhook-configurations.namespace=all
    validating-webhook-configurations.selector=app.kubernetes.io/name=openservicemesh.io
    validating-webhook-configurations.output=json
---
apiVersion: v1 
kind: ConfigMap 
metadata:
  name: servicemesh-config
  namespace: aks-periscope
data:
  DIAGNOSTIC_SERVICEMESH_LIST: osm-controller                              # new collector type, collector defines logical steps and consumes configmap 
  
  osm-controller.servicemesh.properties: |                                 #prefix matches the name of one of the entries in DIAGNOSTIC_SERVICEMESH_LIST
    namespaces.selector=openservicemesh.io/monitored-by={{meshName}}
  
  osm-controller.diagnostic-kubeobjects.properties |                       #servicemesh collector runs the logic for gathering the list of meshes, and list of namespaces that each mesh is monitoring (using above servicemesh.properties.namespaces.selector) 
    metadata.kubeobject=namespaces                                         #then for each namespace it runs this collection of "kubeobjects" collectors 
    metadata.selector=jsonpath={..metadata}
    metadata.output=json
    
    services_table.kubeobject=services
    services_configs.kubeobject=services
    services_configs.output=json
    
    endpoints_table.kubeobject=endpoints
    endpoints_configs.kubeobject=endpoints
    endpoints_configs.output=json
    
    configmaps_table.kubeobject=configmaps
    configmaps_configs.kubeobject=configmaps
    configmaps_configs.output=json
    
    ingresses_table.kubeobject=ingresses
    ingresses_configs.kubeobject=ingresses
    ingresses_configs.output=json
  
    service_accounts_table.kubeobject=serviceaccounts
    service_accounts_configs.kubeobject=serviceaccounts
    service_accounts_configs.output=json
  
  osm-controller-podLogs.properties |                                                     #podLogs could be implemented as either (1) its own discrete collector or (2)baked as part of servicemeshcollector  
    pod.selector="custom-columns=NAME:{..metadata.name},NAMESPACE:{..metadata.namespace}' #perhaps this is always the same and so we can drop it from the yaml and just implement as default behaviour - only applicable if going with (2) above
---

Reply:
@Tatsinnit @itowlson any thoughts on the proposal above?

Reply:
@davidkydd I don't really know enough to comment in this particular case, sorry.  But if you see a use for chained collectors in the near term, then it's probably worth designing for it as long as it's not onerous.

Reply:
Thanks for eyeballing anyway @itowlson, gives me confidence it's at least comprehensible. There are other pending feature requests (Arc team) so maybe we see what their requirements look like before making the call. I *think* chained collectors would be a generally useful feature but I think all kinds of stupid stuff so concrete examples help

Reply:
I will add my 2 cents as well (before discussion moves on), I can only do full review once I have time sometime tonight:

I think its great we are discussing that aspect, but I am not entirely sure how onerous it will be, in case you have some architecture in mind then please feel free to share in a draft PR for us to dig deeper and take a look, please?

Essentially if OSM and ARC work are 2 different implementation with current architecture in place, I will recommend to treat them differently, also it is important to note, that the feature request fro ARC will eventually be a PR from ARC team just like how OSM team is doing it. We will follow similar mechanism like we do in various AKS extension et. al. that internal teams owns the implementation of their mods in the tool, we will make sure it is maintainable but the will own this.

I would encourage you to share across the what chainable collector idea you have in draft and we can see how viable it will be and how different with current architecture it will be? (new can discuss that in there)

Key things to keep in mind are:

* Are we introducing something which will become hard to maintain? (aka will be great if you share your draft implementation with the idea)
* Is current design or architecture has any cons hence the suggestions is made?
* I would keep separation of concerns intact so that the new features from any team follow a specific pattern and don't break any exist functionality.

Reply:
@davidkydd +1, concrete examples will be _absolutely invaluable_ in generalising

Reply:
and +1 to all of @Tatsinnit's observations and caveats as well

Reply:
@Tatsinnit sorry I don't have any draft implementations or anything yet, just an idea at very early stages. It was prompted by the recognition that what the code in this PR is fundamentally doing is:

1. running a bunch of kubeobject collectors that also include a selector and output string (osmlogs.collector.go line 44-55)
2. running a kubeobject collector for deployments with a selector and output string (osmlogs.collector.go line 32)
2.1 for each deployment.meshName returned by the kubeobject collector in 2., run another kubeobject collector to get all the namespaces that match a selector incorporating the deployment.meshName (osmlogs.collector.go line 63)
2.1.1 for each namespace returned in 2.1, unioned with the namespaces output by another kubeobject collecter (osmlogs.collector.go line 57 & 67), run a group of kubeobject collectors to get the various types of kubeobjects we are interested in from these namespaces (the collectDataFromNamespaces func).

If, rather than doing all of this using loops in the go code, we implemented a way to chain kubeobject collectors together (so that the output of the kubeobject in 2. could act as part of the input for the kubeobject in 2.1), we could define the nesting of collectors declaratively in the yaml itself, e.g. something like:

---
apiVersion: v1 
kind: ConfigMap 
metadata:
  name: kubeobject-config
  namespace: aks-periscope
data:
    first-collector.properties |                                                   #first collector is equivalent to step 2. above
            first-collector.kubeobject=deployments                                        
            first-collector.selector=app=osm-controller
            first-collector.output=jsonpath={..meshName}
     second-collector.properties |                                            #second collector is equivalent to step 2.1 above
            dependson=first-collector
            second-collector.kubeobject=namespaces                                        
            second-collector.selector=openservicemesh.io/monitored-by="+first-collector.output
            second-collector.output=jsonpath={..name}
     third-collector.properties |                                                #third collector is equivalent to step 2.1.1 above
            dependson=second-collector
            third-collector.kubeobject=services
            third-collector.namespace=second-collector.output    
            ...... etc for each resource in collectDateFromNamespaces func, all of which depend on second-collector
            
The intended benefit is that more complicated use cases of Periscope (such as what is contained in this PR) could be supported by the building blocks of Periscope itself; and rather than requiring code changes and a whole new build and container image to be pushed and referenced, the same thing can be accomplished with just changes to the deployment yaml.

Note I am also _not_ suggesting we implement chainable collectors right now, its just an idea that seems to be a good fit for the current scenario, as well as a generally useful capability going forward (premature generalization is a thing!).

What I _am_ suggesting we do is extend the kubeobject collector just a little bit, to support selector and output strings, and to be defined as a configmap with file-like keys. Once this is in place it should be straightforward to refactor the code in this PR to implement a more generalized ServiceMeshCollector, and the OSM specific parts can be moved out to the yaml where they belong.
    
It would also be worth taking a look at a default istio example to see if this has the same general shape as OSM (in terms of the k8s object relationships that would be baked into the collector). Otherwise our ServiceMeshCollector might not be very generalizable after all! :)

Hope that make sense? Hit me up on teams or something if you prefer once you've had time to look and we can chat more.

Reply:
@davidkydd , Take your time please and  no-rush for your Draft PR, like mentioned as this could be a great idea and we could explore this with concrete example as mentioned in comments above, that way we can explore all the key mention in here  https://github.com/Azure/aks-periscope/pull/48#discussion_r623554119  **and**  https://github.com/Azure/aks-periscope/pull/48#discussion_r623550397 

Doing your suggestions will help us to explore the idea and design impact from what current design is and it as separate draft PR help us to share thoughts and concrete generalisation example will help us to better understand. Thank you. 🙏

Reply:
Sweet as, I'll have a go at the extensions to the kubeobject collector first then. @SanyaKochhar I assume there isn't a pressing need to have your current PR merged (as you can deploy a periscope build out of a private ACR)?

Reply:
Hi @davidkydd! It would be great to augment Periscope with these changes sooner than later - Open Service Mesh is already in public preview as on AKS addon.

Once we have the prototype in this PR working, it should be straight forward to refactor to DRY it up, and later refactor again to fit the new design you are working on.

We could also wrap these changes with a feature flag to be able to continue to iterate with smaller PRs.

I propose we
1. address as many of the immediate comments to this PR as possible to make it fit the current design
2. merge the PR
3. adapt OSM + SMI code to the new model / design you are working on

Do these sound like reasonable next steps?

Reply:
Hi @draychev :) sure, in that case refactoring later should be fine. 

Feature flag at a minimum would be great as test coverage is currently very limited so regressions are significant concern. Should be able to implement one by adding a new envFrom configmap defining which collectors are enabled (simple list of strings based on type names in collector.go should suffice), and then reading the config and conditionally adding only those collectors that are enabled in aks-periscope.go. Open to better ideas though :)

Other than that I believe the only blocker to merging is the removal of the aks-periscope binary at the project root. We then manually build and push a new containerImage to aksrepos.azurecr before updating the image tag ourselves in a separate PR - is that the right process @Tatsinnit?

Reply:
Agree:

* This is the best step moving forward, and like the idea of treating any design refactor separately so that we can actually explore what the idea is in more clarity.

* I will add bunch of checklist below as separate comment and yes, at least basic checks and due-diligence to check that these changes work with OSM\SMI feature flagged and don't interfere with normal workflow.
     * Key thing is this PR **Can-Only** be merged once all those checks are satisfied. So, please about eager merges.
     * At Dev-level falls will help this code path to only execute when the conditions are met. 

* Cool, @davidkydd  As mentioned, I would recommend you can open a workitem to track this specific re-design part and self assign it to yourself please, for the thoughts and ideas you've shared for this design refactor separately, and I believe this no longer is part of this drop, so this is un-blocked now, I think if we all agree then we could mark this resolved once we have that work item open or I could do it latter and assign it to you. 
    * When re-design \ refactoring the idea: please do take care ofpoint mentioned here  https://github.com/Azure/aks-periscope/pull/48#discussion_r623554119  **and**  https://github.com/Azure/aks-periscope/pull/48#discussion_r623550397 

Please feel free to add anything else need discussion otherwise this kind of sums it all.

Thank you guys 🙏